### PR TITLE
fork from rbenv rather than copy?

### PR DESCRIPTION
**Pull request by jasonkarns**
I would have liked to have seen nodenv (+ node-build and plugins) forked from rbenv rather than copied. If they shared a common commit history, it would have made it trivial to merge in any 'upstream' changes.

I wouldn't mind doing the work to graft the history back in, if it would be of interest. Personally, I would feel much more comfortable with the stability of nodenv if I knew that any bugfixes/security fixes/performance improvements that get made to rbenv were easily applied to nodenv as well. Especially since rbenv (at least presently) has a much higher user base and active contributors.

Would this be of interest?

### CODE REVIEW

**Comment by jasonkarns** [comment]
As you can see, I've converted this Issue into a PR from my [fork of nodenv](https://github.com/jasonkarns/nodenv).

My fork has all the history of rbenv grafted in, while maintaining nodenv's history, (and maintaining the original initial commit from nodenv to still allow force-pushing to GitHub).

Steps I took:

Ideally, I would have grafted rbenv's history into nodenv by giving nodenv's initial commit (019247c) a parent from rbenv's history. However, this would change the initial commit's sha, which would prevent force-pushing to GitHub. So instead, I grafted in rbenv's history to nodenv's second commit (originally 2cee5e9). I chose ea3203d from rbenv as that was master's HEAD at the time nodenv was created. After grafting that in, I ran `filter-branch` to make the change permanent.

At this point, I now had nodenv sharing history with rbenv! But there was one more graft that was needed. In April of 2014, nodenv was updated with the latest (at the time) changes from rbenv. So I took that commit (originally 956a73b) and grafted in its *spirit* parent from rbenv (f71e227). And ran `filter-branch` again.

Now nodenv shares rbenv's history

To summarize:
- jasonkarns/master is content-identical to nodenv/master
- jasonkarns/master shares rbenv's history and has the appropriate `merge-base` (f71e227), which means future merges from upstream rbenv will be as simple as `git merge rbenv/master`
- jasonkarns/master retained the initial nodenv commit so it can still be force-pushed to github

Recommendation: rather than 'merge' this PR, @OiNutter will need to add my fork as a remote, and reset master to jasonkarns/master, followed by a force-push to github. This way all the history modification I did is preserved without showing an even more complicated merge history.

**Comment by jasonkarns** [comment]
As a discussion point, it's also worth deciding if we want this history to be as 'correct in spirit' as possible. (Which would mean using my fork as described above.) Or if all we care about is having the correct `merge-base` between nodenv and rbenv. If the latter, then simply doing `git merge -s ours rbenv/f71e22768c8bc02a10fc917f05072fdc68d71be4` into nodenv/master would suffice.

Doing just the merge would pull in rbenv's history, maintain nodenv's history (without rewriting it!) and still result in the proper merge-base for future merges. The downside being that the history is not representative of how the code actually came to be.

**Comment by ianks** [comment]
I agree with this idea in spirit. In fact, so much so that I think this work would be great to see in upstream rbenv. This way, we could decouple the architecture from binary formats.

**Comment by jawshooah** [comment]
Just to weigh in, I think that having the correct merge-base is sufficient. Should we need to dig deeper into the history, it's simple enough to stroll on over to rbenv.

**Comment by jasonkarns** [comment]
@jawshooah I think I'm beginning to prefer that as well. It's less intrusive on the existing history of this (and the other plugins) repo. Closing this since we don't want this PR merged.

**Comment by jasonkarns** [comment]
Alternative PR submitted that does a merge from rbenv/master while ignoring rbenv's code: #14

**Comment by peterkc** [comment]
Thanks for cleaning up the issues @jasonkarns.

From what I can see in the main rbenv codebase, the majority of it is very generic and can be repurpose. Most of the work for nodeenv would be in the custom builder [node-build](https://github.com/OiNutter/node-build)
